debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java

```diff
@@ -0,0 +1,303 @@
+                close();
+            }
+        }
+
+        private void fetchEvents(MongoChangeStreamCursor<ChangeStreamDocument<TResult>> cursor) throws InterruptedException {
+            while (isRunning()) {
+                var beforeEventPollTime = clock.currentTimeAsInstant();
+                var event = new ResumableChangeStreamEvent<>(cursor.next(), cursor);
+                event.document.ifPresent(doc -> LOGGER.trace("Polled Change Stream event: {}", event));
+                metrics.onSourceEventPolled(event, clock, beforeEventPollTime);
+
+                if (event.hasDocument()) {
+                    enqueue(event);
+                }
+                else {
+                    throttleIfNeeded(event);
+                }
+            }
+        }
+
+        private void throttleIfNeeded(ResumableChangeStreamEvent<TResult> event) throws InterruptedException {
+            noMessageIterations++;
+            if (noMessageIterations >= THROTTLE_NO_MESSAGE_BEFORE_PAUSE) {
+                enqueue(event);
+                noMessageIterations = 0;
+                pause.sleepWhen(true);
+            }
+        }
+
+        private void enqueue(ResumableChangeStreamEvent<TResult> event) throws InterruptedException {
+            this.capacity.acquire();
+            queue.offer(event);
+        }
+    }
+
+    public static <TResult> BufferingChangeStreamCursor<TResult> fromIterable(
+                                                                              ChangeStreamIterable<TResult> stream,
+                                                                              MongoDbTaskContext taskContext,
+                                                                              MongoDbStreamingChangeEventSourceMetrics metrics,
+                                                                              Clock clock) {
+        var pause = DelayStrategy.constant(taskContext.getConnectorConfig().getPollInterval());
+        var fetcher = new EventFetcher<>(stream, 1000, pause, metrics, clock);
+        var executor = Threads.newFixedThreadPool(MongoDbConnector.class, taskContext.serverName(), "replicator-buffer", 1);
+
+        return new BufferingChangeStreamCursor<>(fetcher, executor);
+    }
+
+    /**
+     * Creates new prefetching cursor
+     *
+     * @param fetcher MongoDB change event fetcher
+     * @param executor executor used to dispatch buffering thread
+     */
+    BufferingChangeStreamCursor(EventFetcher<TResult> fetcher, ExecutorService executor) {
+        this.fetcher = fetcher;
+        this.executor = executor;
+    }
+
+    public BufferingChangeStreamCursor<TResult> start() {
+        executor.submit(fetcher);
+        return this;
+    }
+
+    @Override
+    public ResumableChangeStreamEvent<TResult> tryNext() {
+        var event = fetcher.poll();
+        if (event != null) {
+            lastResumeToken = event.resumeToken;
+        }
+        return event;
+    }
```

Comment:
@jpechane  if we want to add the throttling here, how would you envision it?  Say this method is called and no event is available. It doesn't make sense, IMHO, to try N-times and then pause before returning null.  Maybe count the number of empty polls on the class level and if the N previous calls return null we pause before trying again?

Reply:
Or maybe a progressive pause would be better here, starting with a small one? If it works nicely we could probably adopt it for the main loop too.

Reply:
OK. I will look into it. By main loop I suppose you mean the one in the fetcher?

Reply:
Sorry, yes